File: lima_linguisticdata/scripts/flex.pl
#!/usr/bin/perl -W
#   Copyright 2002-2013 CEA LIST
#    
#   This file is part of LIMA.
#
#   LIMA is free software: you can redistribute it and/or modify
#   it under the terms of the GNU Affero General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   LIMA is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU Affero General Public License for more details.
#
#   You should have received a copy of the GNU Affero General Public License
#   along with LIMA.  If not, see <http://www.gnu.org/licenses/>
use strict;
use utf8;

my ($deffile, $infile, $workdir, $outfile, $excludesfile, $debug) = @ARGV;
open (OUTFILE,">:utf8","$workdir/$outfile") || die "Cannot create $outfile\n";
open (FILETRACE,">:utf8","$workdir/$infile.log") || die "Cannot create $infile.log\n";

(defined $debug) || ($debug=0);

my %categstypes;
my %struct;
my %excludes;

loadData($deffile, \%categstypes, \%struct);

if (defined $excludesfile) {
    loadExcludes($excludesfile, \%excludes);
}
else {
    print STDERR "no excludes file\n";
}

# traitement des formes
print ("Traitement du fichier des mots simples\n");
open (INFILE,"<:utf8",$infile) || die "Cannot open mots-simples.txt\n";

my $oldfh = select(STDOUT);
$| = 1;
select($oldfh);

my $infileLineNum = 0;
while (<INFILE>)
{

  $infileLineNum++;
  if ($infileLineNum%1000 == 0)
  {
	  print "\n$infileLineNum";
  }
  chomp;
  next if ($_ eq "" || (/^#/) );
  my $line = $_;
  if ($_ !~ /^([^;]*);([^;]*);([^;]*);([^;]*);([^;]*)/)
  {
    print FILETRACE "Malformed line $infileLineNum: $_\n";
    next;
  } 
  my $lem = $3;
  my $categ = $4;
  my $norm = $5;
  # faire un miroir du lemme
  my $leminv = "";
  for (my $i=length($lem)-1;$i>=0;$i=$i-1) 
  {
      $leminv=$leminv . substr($lem,$i,1);
  }
  # transcoder la vg pour macrocategorie et prendre les autres informations
  # traitement des noms a ne pas flechir
#  if ($lem=~/-/)    # Greg: on va traiter les mots avec tirets 4/1/05
#  {
#    print OUTFILE ($lem,";",$categ,";;;",$lem,";",$norm,";\n");
#  }
#  elsif ($lem=~/[A-Z].*[A-Z]/) 
  if ($lem=~/[A-Z].*[A-Z]/)
  {
    &printOutput($lem.";".$categ.";;;".$lem.";".$norm.";");
  }
  else
  {    
    traitement(\%struct, $line, $categstypes{$categ}, $lem, $leminv, $categ, $norm);
  }
}
print "\n";

sub traitement
{
  my ($struct, $line, $types, $lem, $leminv, $categ, $norm) = @_;
  if (!defined $types)
  {
    &printOutput($lem.";".$categ.";;;".$lem.";".$norm.";");
    return;
  }

  my @types = @$types;
  foreach my $type (@types)
  {
    my ($models, $modelsdata, $modelstab);
    $models = \%{${$struct{$type}}{"models"}};
    $modelstab = \@{${$struct{$type}}{"modelstab"}};
    $modelsdata = \%{${$struct{$type}}{"modelsdata"}};
    
    my $model;
    my $commun;
    my $allDone;

    my $subleminv;


    $subleminv=$leminv."\$"; 
    $allDone=0; 

    while($allDone==0) {
      $debug && print "subleminv='$subleminv'\n";
    if (defined ($$models{$subleminv})) 
      {
        $model = $$models{$subleminv};
        $commun = $subleminv;
        last;
      }
      if(length($subleminv)>0) {
            chop ($subleminv);
	}
      else { $allDone=1; }
  }

#    foreach my $masque (@$modelstab)
#    {
#      $debug && print "masque='$masque'\n";
#      if ($leminv=~/^$masque/)
#      {
#        $model = $$models{$masque};
#        $commun = $masque;
#        last;
#      }
#    }

    if (!defined($model)) 
    {
      print FILETRACE "Unable to handle line (no model found) $infileLineNum: $line\n";
      return;
    }
    
    $debug && print "$type: lemme='$lem' leminv='$leminv' model='$model' commun='$commun'\n";
    if (!defined($$modelsdata{$model})) 
    {
      print FILETRACE "Unable to handle line (no data for model $model) $infileLineNum: $line\n";
      return;
    }
    my ($racine, $longueur);
    racine($type, \$racine,\$longueur,$model,$lem,$commun);
    foreach my $map ( @{$$modelsdata{$model}} ) 
    {
      my %map = %{$map};
      if (length($$map{"forme"}) > 0)
      {
        #on cherche la partie a conserver du lemme en comparant la longueur mot modifie a celui de fin
        #la difference est la longueur a garder.
        #pour chaque ligne ou le mot modifie est present on construit la forme en prenant la partie fixe et en ajouta$adjectifnt la forme modifiee amputee de sa 
        #forme fixe
        #ecrire dans le fichier de sortie la forme, la vg , les proprietes, le lemme  
        my $forme1=$$map{"forme"};
        my $third;
        my $fourth;
        if ($type eq "verbe")
        {
          $third = $$map{"third"};
          $fourth = $$map{"fourth"};
        }
        else
        {
          $fourth = $$map{"third"};
          $third = $$map{"fourth"};
        }
        # traitement des variantes separees par "/" ( attention une seule variante possible)
        my $forme_sortie;
        if ($$map{"forme"}=~/\//) 
        {
          my @table=split("\/",$$map{"forme"});
          
          $forme_sortie=$racine . substr($table[0],$longueur);
          my $outline = "$forme_sortie;$categ;$third;$fourth;$lem;$norm;";
          $outline =~ s/\?//g;
          &printOutput($outline);
          $forme1=$table[1];
        }
        $forme_sortie = $racine . substr($forme1,$longueur);
#        $debug && print "forme1 = '$forme1' ; forme_sortie = '$forme_sortie'\n";
        my $outline = "$forme_sortie;$categ;$third;$fourth;$lem;$norm;";
        $outline =~ s/\?//g;
        &printOutput($outline);
      }
    }
  }
}


sub racine
{
  my ($type, $racine,$longueur,$model,$lem,$commun) = @_;
  
  if ($type eq "verbe")
  {
    my $llem=length($lem)-1;
    my $lmodel=length($model)-1;
    while (substr($lem,$llem) eq substr($model,$lmodel) && $llem >= 0 && $lmodel >=0) 
    { 
      $llem--; $lmodel--;
    }
    $$racine=substr($lem,0,$llem+1);
    $$longueur = $lmodel+1;
  }
  else
  {
    my $lfin=length($commun); 
    if (chop($commun) eq "\$") 
    {
      $lfin=$lfin-1;
    }
    $$racine=substr($lem,0,length($lem)-$lfin);
    $$longueur=length($model)-$lfin;
    if ($model=~/\^/) 
    {
      $$longueur--;
    }
  }
  $debug && print "Apres racine(): racine='$$racine' longueur='$$longueur'\n"
}

sub loadData
{
  my ($file, $categsTypes, $struct) = @_;

  open (DEFFILE,"<:utf8", $file) || die "Cannot open $file\n";
  
  my $lineNum = 0;
  while (<DEFFILE>)
  {
    chomp;
    $lineNum++;
    if ($_ eq "" || (/^#/) ) {next ;}
    die "Malformed definition line $lineNum in $file: $_\n" unless (/^[^;]*;[^;]*;[^;]*;[^;]*\s*$/);
    my ($type, $categs, $ficmodel, $fictable) = split(";", $_);
    my @categs = split(",", $categs);
    foreach my $categ (@categs)
    {
      if (! defined $$categsTypes{$categ} )
      {
        my @types;
        $$categsTypes{$categ} = \@types;
      }
      push @{$$categsTypes{$categ}}, $type;
    }
    my (@modelentries, %models, @modelstab);

    my $nbmodel = loadModels($ficmodel, \%models, \@modelstab);
    print "Got $nbmodel $type modelword\n";

    my (@models,@third,@fourth,%direct,%modelsdata);

    my $nbtable = loadTable($type, $fictable, \%modelsdata);
    print "Got $nbtable $type table elems\n";
    
    my %nomtabTab;
    $nomtabTab{"models"} = \%models;
    $nomtabTab{"modelsdata"} = \%modelsdata;
    $nomtabTab{"modelstab"} = \@modelstab;
  
    $$struct{$type} = \%nomtabTab;
  }
  close DEFFILE;
}

sub loadModels
{
  my ($dicfile, $models, $modelstab) = @_;
  print ("Chargement de $dicfile\n");
  open (FILEDIC,"<:utf8", $dicfile) || die "Cannot open $dicfile\n";
  
  my $lineNum = 0;
  while (<FILEDIC>)
  {
    chomp;
    if ($_ eq "" || (/^#/) ) {next;}
    $lineNum++;
    die "Malformed model line $lineNum in $dicfile: $_\n" unless (/^([^;]*);([^;]*);\s*$/);
    
    push @$modelstab, $1;
    $$models{$1} = $2;
  }
  close (FILEDIC);
  return $lineNum;
}
    
sub loadTable
{
  # third = temps pour verbe, genre pour autres
  # fourth = personne pour verbe, nombre pour autres
  my ($type, $tablefile, $modelsdata) = @_;
  print ("Chargement de $tablefile\n");
  open (FILETABLE,"<:utf8", $tablefile) || die "Cannot open $tablefile\n";

  my $oldmodel = "";
  my $lineNum = 0;
  my $ok = 1;
  my @maps; 
  while (<FILETABLE>) 
  {
    chomp;
    if ($_ eq "" || (/^#/) ) {next;}
    $lineNum++;
    if ($_ !~ /^([^;]*);([^;]*);([^;]*);([^;]*)$/)
    {
      print STDERR "Malformed table line $lineNum in $tablefile: $_\n";
      $ok = 0;
      next;
    }
    my $modelword = $1 ;
    my $formes = $2;
    my $thirds = $3;
    my $fourths = "";
    if (/^([^;]*);([^;]*);([^;]*);([^;]+;?)\s*$/)
    {
      if ($type ne "verbe")
      {
        print STDERR "Malformed non-verb table line $lineNum in $tablefile (should have 3 columns):\n$_\n\n";
        $ok = 0;
        next;
      }
      $fourths = $4;
    }
#    push @$models, $modelword;
    if ($oldmodel ne $modelword && $oldmodel ne "")
    {
      my @modelwordmaps = @maps;
      $debug && print "fixing modelsdata{$oldmodel}\n";
      $$modelsdata{$oldmodel} = \@modelwordmaps;
      @maps = ();
    }
    $oldmodel = $modelword;
    my %map;
    $map{"forme"} = $formes;
    $map{"third"} = $thirds;
    $map{"fourth"} = $fourths;
    $debug && print "forme='$formes' third='$thirds' fourth='$fourths'\n\n";
    push @maps, \%map;
  }
  my @modelwordmaps = @maps;
  $debug && print "fixing modelsdata{$oldmodel}\n";
  $$modelsdata{$oldmodel} = \@modelwordmaps;
  close (FILETABLE);
  die "Had problems reading $tablefile\n"  unless ($ok);
  return $lineNum;
}

sub loadExcludes
{
  my ($excludesfile, $excludes) = @_;
  print ("Chargement du fichier d'exclusions $excludesfile\n");
  if ( -f $excludesfile )
  {
    open (EXCLUDES,"<:utf8", $excludesfile) 
        || die "Cannot open $excludesfile\n";
    while (<EXCLUDES>)
    {
      chomp;
      $$excludes{$_}=1;
    }
    close EXCLUDES;
  }
  else {
      print STDERR "file $excludesfile not found\n";
  }
}

sub printOutput {
    my ($line)= @_;
    if ( ! exists $excludes{$line} )
    {
        print OUTFILE ("$line\n");
    }
    else 
    {
        print FILETRACE "Excluding line: $line\n";
    }
}

File: lima_linguisticdata/scripts/flex.py
import argparse
import codecs
import os
import sys

from tqdm import tqdm

output_file, filetrace = None, None
excludes = {}


def load_data(file, categs_types, struct):
    with codecs.open(file, 'r', 'utf-8') as deffile:
        for line_num, line in enumerate(deffile, 1):
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            if not line.count(';') == 3:
                sys.exit(f"Malformed definition line {line_num} "
                         f"in {file}: {line}")

            type_, categs, ficmodel, fictable = line.split(';')
            categ_list = categs.split(',')

            for categ in categ_list:
                if categ not in categs_types:
                    categs_types[categ] = []
                categs_types[categ].append(type_)

            models, modelstab = {}, []
            nbmodel = load_models(ficmodel, models, modelstab)
            print(f"Got {nbmodel} {type_} modelword")

            modelsdata = {}
            nbtable = load_table(type_, fictable, modelsdata)
            print(f"Got {nbtable} {type_} table elems")

            struct[type_] = {"models": models,
                             "modelsdata": modelsdata,
                             "modelstab": modelstab}


def load_models(dicfile, models, modelstab):
    print(f"Loading {dicfile}")
    with codecs.open(dicfile, 'r', 'utf-8') as filedic:
        for line_num, line in enumerate(filedic, 1):
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            if not line.count(';') == 2:
                sys.exit(f"Malformed model line {line_num} "
                         f"in {dicfile}: {line}")

            model, val, _ = line.split(';')
            modelstab.append(model)
            models[model] = val

    return line_num


def load_table(type_, tablefile, modelsdata):
    print(f"Loading {tablefile}")
    oldmodel, maps = "", []
    ok = True
    with codecs.open(tablefile, 'r', 'utf-8') as filetable:
        for line_num, line in enumerate(filetable, 1):
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            parts = line.split(';')
            if len(parts) < 3 or len(parts) > 4:
                print(f"Malformed table line {line_num} "
                      f"in {tablefile}: {line}",
                      file=sys.stderr)
                ok = False
                continue
            if type_ != "verbe" and len(parts) == 4:
                print(f"Malformed non-verb table line {line_num} "
                      f"in {tablefile} (should have 3 columns):\n{line}",
                      file=sys.stderr)
                ok = False
                continue

            modelword, formes, thirds = parts[0], parts[1], parts[2]
            fourths = parts[3] if len(parts) == 4 else ""

            if oldmodel and oldmodel != modelword:
                modelsdata[oldmodel] = maps[:]
                maps.clear()

            oldmodel = modelword
            maps.append({"forme": formes, "third": thirds, "fourth": fourths})

        if oldmodel:
            modelsdata[oldmodel] = maps[:]
    if not ok:
        sys.exit(f"Had problems reading {tablefile}")
    return line_num


def load_excludes(excludesfile, excludes):
    print(f"Loading exclusions file {excludesfile}")
    if os.path.isfile(excludesfile):
        with codecs.open(excludesfile, 'r', 'utf-8') as excludefile:
            for line in excludefile:
                line = line.strip()
                excludes[line] = True
    else:
        print(f"File {excludesfile} not found", file=sys.stderr)


def racine(type_, model, lem, commun):
    if type_ == "verbe":
        llem, lmodel = len(lem) - 1, len(model) - 1
        while llem >= 0 and lmodel >= 0 and lem[:llem] == model[:lmodel]:
            llem -= 1
            lmodel -= 1
        return lem[:llem + 1], lmodel + 1

    lfin = len(commun) - (1 if commun.endswith('$') else 0)
    racine = lem[:len(lem) - lfin]
    longueur = len(model) - lfin
    if '^' in model:
        longueur -= 1

    return racine, longueur


def traitement(struct, line, types, lem, leminv, categ, norm, filetrace):
    if types is None:
        print_output(f"{lem};{categ};;;{lem};{norm};")
        return

    for type_ in types:
        models, modelsdata = (struct[type_]['models'],
                              struct[type_]['modelsdata'])
        subleminv = leminv + '$'
        model, commun = None, None

        while subleminv:
            if subleminv in models:
                model = models[subleminv]
                commun = subleminv
                break
            subleminv = subleminv[:-1]

        if model is None:
            filetrace.write(f"Unable to handle line (no model found) {line}\n")
            return

        if model not in modelsdata:
            filetrace.write(f"Unable to handle line "
                            f"(no data for model {model}) {line}\n")
            return

        racine_val, longueur = racine(type_, model, lem, commun)
        for map_ in modelsdata[model]:
            forme1 = map_['forme']
            third = map_['third'] if type_ == 'verbe' else map_['fourth']
            fourth = map_['fourth'] if type_ == 'verbe' else map_['third']

            forme_sortie = racine_val + forme1[longueur:]
            if '/' in map_['forme']:
                table = map_['forme'].split('/')
                forme_sortie = racine_val[table[0]:table[0]+longueur]
                outline = (f"{forme_sortie};{categ};{third}"
                           f";{fourth};{lem};{norm};")
                outline = outline.replace('?', '')
                print_output(outline)
                forme1 = table[1]

            forme_sortie = racine_val[forme1:forme1+longueur]
            outline = f"{forme_sortie};{categ};{third};{fourth};{lem};{norm};"
            outline = outline.replace('?', '')
            print_output(outline)


def print_output(line):
    global output_file, filetrace, excludes
    if line not in excludes:
        output_file.write(f"{line}\n")
    else:
        filetrace.write(f"Excluding line: {line}\n")


def main():
    parser = argparse.ArgumentParser(description="Process some files.")
    parser.add_argument('deffile', help="The definition file")
    parser.add_argument('infile', help="The input file")
    parser.add_argument('workdir', help="The working directory")
    parser.add_argument('outfile', help="The output file")
    parser.add_argument('--excludesfile', help="The excludes file")

    args = parser.parse_args()

    global output_file, filetrace, excludes
    output_file = codecs.open(
      os.path.join(args.workdir, args.outfile), 'w', 'utf-8')
    filetrace = codecs.open(
      os.path.join(args.workdir, f"{args.infile}.log"), 'w', 'utf-8')

    categs_types, struct = {}, {}

    load_data(args.deffile, categs_types, struct)

    if args.excludesfile:
        load_excludes(args.excludesfile, excludes)
    else:
        print("no excludes file", file=sys.stderr)

    print("Processing simple words file")
    with codecs.open(args.infile, 'r', 'utf-8') as infile:
        for line_num, line in enumerate(
          tqdm(infile, desc="Flexing lemmas"), 1):
            line = line.strip()
            if not line or line.startswith('#'):
                continue
            if len(line.split(';')) < 5:
                filetrace.write(f"Malformed line {line_num}: {line}\n")
                continue

            parts = line.split(';')
            lem, categ, norm = parts[2], parts[3], parts[4]
            leminv = lem[::-1]

            if sum(1 for char in lem if char.isupper()) >= 2:
                print_output(f"{lem};{categ};;;{lem};{norm};")
            else:
                traitement(struct, line, categs_types.get(categ), lem,
                           leminv, categ, norm, filetrace)

    output_file.close()
    filetrace.close()


if __name__ == "__main__":
    main()
